### PR TITLE
Do not use TLS for replication devices as calls into them from autogr…

### PR DESCRIPTION
**Pull request by dlibenzi**
…ad are coming from "alien" threads.

### CODE REVIEW

File: third_party/xla_client/xrt_computation_client.cc [hunk 1]
```diff
@@ -1410,13 +1408,15 @@ std::vector<std::string> XrtComputationClient::GetAllDevices() const {
...
+  replication_devices_ = std::move(devices);
 }
 
-const std::vector<std::string>& XrtComputationClient::GetReplicationDevices()
-    const {
-  return g_replication_devices;
+std::shared_ptr<std::vector<std::string>>
+XrtComputationClient::GetReplicationDevices() {
+  std::lock_guard<std::mutex> lock(lock_);
+  return replication_devices_;
```
Comment (JackCaoG) [inline third_party/xla_client/xrt_computation_client.cc:1419]: Does this means all thread share the same `replication_devices_`?

Comment (dlibenzi) [reply third_party/xla_client/xrt_computation_client.cc:1419]: Yes, that is always the case.
The replication devices is a global concept, and I don't even remember why it was TLS in the first place.

**Comment by dlibenzi** [comment]
> Thanks!!
> Btw is it worth porting back to r1.6 branch as well?

This might be worth a back-port. It will likely be a non trivial back-port as it touched code which was not there in 1.6.

**Comment by matt-peters** [comment]
Will do.  Thanks for the super quick fix!

**Pull request merged**